File: Web/Scripts/saves/savemanager.py
# Fyleria Engine
# Copyright © 2018 Go Go Gecko Productions

# Imports
import os, sys, json
import GeckoUtility
import GeckoSaves
from GeckoModule import StoreRunResult

# Get managers
g_SaveManager = GeckoSaves.GetSaveManager()

#################################################################################################
## Load save
def LoadSave(uSlot, sSaveJson):
    g_SaveManager.LoadSave(uSlot, GeckoSaves.GetSaveFromJsonString(sSaveJson))
#################################################################################################
## Create save
def CreateSave(uSlot):
    g_SaveManager.CreateSave(uSlot)
#################################################################################################
## Unload save
def UnloadSave(uSlot):
    g_SaveManager.UnloadSave(uSlot)
#################################################################################################
## Unload all saves
def UnloadAllSaves():
    g_SaveManager.UnloadAllSaves()
#################################################################################################
## Does save exist
def DoesSaveExist(uSlot):
    result = g_SaveManager.DoesSaveExist(uSlot)
    StoreRunResult(json.dumps(result))
#################################################################################################
## Get save capacity
def GetSaveCapacity():
    result = g_SaveManager.GetSaveCapacity()
    StoreRunResult(json.dumps(result))
#################################################################################################
## Get list of all available save slots
def GetAllAvailableSaveSlots():
    result = g_SaveManager.GetAllAvailableSaveSlots()
    StoreRunResult(json.dumps(result))
#################################################################################################
## Get list of all descriptions from save slots
def GetAllSaveDescriptions():
    result = g_SaveManager.GetAllSaveDescriptions()
    StoreRunResult(json.dumps(result))
#################################################################################################
## Is save capacity reached
def IsSaveCapacityReached():
    result = g_SaveManager.IsSaveCapacityReached()
    StoreRunResult(json.dumps(result))
#################################################################################################
## Get save
def GetSave(uSlot):
    result = g_SaveManager.GetSave(uSlot)
    StoreRunResult(GeckoSaves.ConvertSaveToJsonString(result))
#################################################################################################
## Get all saves
def GetAllSaves():
    result = g_SaveManager.GetAllSaves()
    StoreRunResult(GeckoSaves.ConvertSaveListToJsonString(result))
#################################################################################################
## Pull save from memory
def PullSaveFromMemory(uSlot, sPartyID):
    g_SaveManager.PullSaveFromMemory(uSlot, sPartyID)
#################################################################################################
## Push save into memory
def PushSaveIntoMemory(uSlot):
    g_SaveManager.PushSaveIntoMemory(uSlot)
#################################################################################################
## Save to file
def SaveToFile(uSlot, sFile, sType):
    g_SaveManager.SaveToFile(uSlot, sFile, sType)
#################################################################################################
## Load from file
def LoadFromFile(uSlot, sFile, sType):
    g_SaveManager.LoadFromFile(uSlot, sFile, sType)
#################################################################################################
## Save all to the given directory
def SaveAllToDirectory(sDirectory, sBase, sExt, sType):
    g_SaveManager.SaveAllToDirectory(sDirectory, sBase, sExt, sType)
#################################################################################################
## Load all from the given directory
def LoadAllFromDirectory(sDirectory, sBase, sExt, sType):
    g_SaveManager.LoadAllFromDirectory(sDirectory, sBase, sExt, sType)
#################################################################################################
## Initialize all save slots
def InitializeAllSaveSlots():
    g_SaveManager.InitializeAllSaveSlots()
#################################################################################################
## Initialize empty save slots
def InitializeEmptySaveSlots():
    g_SaveManager.InitializeEmptySaveSlots()
#################################################################################################

File: Web/Project/Libs/Fyleria/Lib/Saves/SaveManager.cpp
// Fyleria Engine
// Copyright © 2016 Go Go Gecko Productions

#include "Saves/SaveManager.h"
#include "Saves/SaveTypes.h"
#include "Utility/Constants.h"

namespace Gecko
{

SaveManager::SaveManager()
    : Singleton<SaveManager>()
{
}

void SaveManager::LoadSave(UByte uSlot, const Save& save)
{
    // Load a save
    m_tSaves[uSlot] = save;
    m_tSaves[uSlot].SetSlot(uSlot);
}

void SaveManager::CreateSave(UByte uSlot)
{
    // Create a new save
    ASSERT_ERROR(!DoesSaveExist(uSlot), "Save slot '%u' was already registered", uSlot);
    Save newSave;
    newSave.SetSlot(uSlot);
    m_tSaves.insert({uSlot, newSave});
}

void SaveManager::UnloadSave(UByte uSlot)
{
    // Unload save
    ASSERT_ERROR(DoesSaveExist(uSlot), "Save slot '%u' was not registered", uSlot);
    m_tSaves.erase(uSlot);
}

void SaveManager::UnloadAllSaves()
{
    // Unload all saves
    m_tSaves.clear();
}

Bool SaveManager::DoesSaveExist(UByte uSlot) const
{
    // Check if save exists
    auto iSearch = m_tSaves.find(uSlot);
    return (iSearch != m_tSaves.end());
}

UByte SaveManager::GetSaveCapacity() const
{
    // Get max number of save slots
    return s_kuMaxSaveSlot;
}

UByteList SaveManager::GetAllAvailableSaveSlots() const
{
    // Get a list of all slots that are not used
    UByteList vSlots;
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        if(!DoesSaveExist(i))
        {
            vSlots.push_back(i);
        }
    }
    return vSlots;
}

StringList SaveManager::GetAllSaveDescriptions() const
{
    // Get all save descriptions
    StringList vDescriptions;
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        vDescriptions.push_back(DoesSaveExist(i) ? GetSave(i).GetDescription() : "");
    }
    return vDescriptions;
}

Bool SaveManager::IsSaveCapacityReached() const
{
    // Get whether save capacity is reached
    STDVector<UByte> vAvailableSlots = GetAllAvailableSaveSlots();
    return vAvailableSlots.empty();
}

Save& SaveManager::GetSave(UByte uSlot)
{
    // Get save
    ASSERT_ERROR(DoesSaveExist(uSlot), "Save slot '%u' was not registered", uSlot);
    return m_tSaves[uSlot];
}

const Save& SaveManager::GetSave(UByte uSlot) const
{
    // Get save
    ASSERT_ERROR(DoesSaveExist(uSlot), "Save slot '%u' was not registered", uSlot);
    auto iSearch = m_tSaves.find(uSlot);
    return iSearch->second;
}

SaveList SaveManager::GetAllSaves() const
{
    SaveList vSaves;
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        vSaves.push_back(GetSave(i));
    }
    return vSaves;
}

void SaveManager::PullSaveFromMemory(UByte uSlot, const IndexedString& sPartyID)
{
    // Get party and attached characters
    CharacterList vCharacters;
    CharacterParty& party = CharacterPartyManager::GetInstance()->GetPartyByID(sPartyID);
    for(const IndexedString& sMemberID : party.GetMembers())
    {
        vCharacters.push_back(CharacterManager::GetInstance()->GetCharacter(sMemberID));
    }

    // After getting party, make sure to regenerate character data
    // We want to have the most accurate data before we save
    party.RegenerateCharacterData();

    // Create save from this
    Save newSave;
    newSave.SetSlot(uSlot);
    newSave.SetTime(party.GetPlayTime());
    newSave.SetParty(party);
    newSave.SetCharacters(vCharacters);
    newSave.SetDescription(party.GetDescription());

    // Load save into manager, potentially overwriting an existing save
    LoadSave(uSlot, newSave);
}

void SaveManager::PushSaveIntoMemory(UByte uSlot)
{
    // Get save from slot
    const Save& save = GetSave(uSlot);

    // Load characters
    for(const Character& character : save.GetCharacters())
    {
        CharacterManager::GetInstance()->LoadCharacter(character.GetCharacterID(), character);
    }

    // Load party
    CharacterPartyManager::GetInstance()->LoadParty(save.GetParty().GetPartyID(), save.GetParty());
}

void SaveManager::SaveToFile(UByte uSlot, const IndexedString& sFile, const IndexedString& sType)
{
    // Save to file based on the file type
    Bool bSuccess = false;
    const FileType eFileType = StringToFileType(sType);
    switch(eFileType)
    {
        case FileType::TextJson:
            bSuccess = GetSave(uSlot).ToFile(sFile);
            ASSERT_ERROR(bSuccess, "Saving to Json file '%s' failed", sFile.c_str());
            break;
        case FileType::BinaryCBOR:
            bSuccess = GetSave(uSlot).ToCBORFile(sFile);
            ASSERT_ERROR(bSuccess, "Saving to CBOR file '%s' failed", sFile.c_str());
            break;
        case FileType::BinaryMsgPack:
            bSuccess = GetSave(uSlot).ToMsgPackFile(sFile);
            ASSERT_ERROR(bSuccess, "Saving to MsgPack file '%s' failed", sFile.c_str());
            break;
        default:
            break;
    }
}

void SaveManager::LoadFromFile(UByte uSlot, const IndexedString& sFile, const IndexedString& sType)
{
    // Load from file based on the file type
    Bool bSuccess = false;
    const FileType eFileType = StringToFileType(sType);
    switch(eFileType)
    {
        case FileType::TextJson:
            bSuccess = GetSave(uSlot).FromFile(sFile);
            ASSERT_ERROR(bSuccess, "Loading from Json file '%s' failed", sFile.c_str());
            break;
        case FileType::BinaryCBOR:
            bSuccess = GetSave(uSlot).FromCBORFile(sFile);
            ASSERT_ERROR(bSuccess, "Loading from CBOR file '%s' failed", sFile.c_str());
            break;
        case FileType::BinaryMsgPack:
            bSuccess = GetSave(uSlot).FromMsgPackFile(sFile);
            ASSERT_ERROR(bSuccess, "Loading from MsgPack file '%s' failed", sFile.c_str());
            break;
        default:
            break;
    }
}

void SaveManager::SaveAllToDirectory(const IndexedString& sDirectory, const IndexedString& sBase, const IndexedString& sExt, const IndexedString& sType)
{
    // Save each slot into a save file
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        FilesystemPath path = FilesystemPath(sDirectory.Get()) / FilesystemPath(sBase.Get() + STDToString(i) + sExt.Get());
        SaveToFile(i, IndexedString(GetNativeFileLocation(path)), sType);
    }
}

void SaveManager::LoadAllFromDirectory(const IndexedString& sDirectory, const IndexedString& sBase, const IndexedString& sExt, const IndexedString& sType)
{
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        FilesystemPath path = FilesystemPath(sDirectory.Get()) / FilesystemPath(sBase.Get() + STDToString(i) + sExt.Get());
        if(DoesFileExist(path))
        {
            LoadFromFile(i, IndexedString(GetNativeFileLocation(path)), sType);
        }
    }
}

void SaveManager::InitializeAllSaveSlots()
{
    // Unload all saves first
    UnloadAllSaves();

    // Create a save for all slots
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        CreateSave(i);
    }
}

void SaveManager::InitializeEmptySaveSlots()
{
    // Create a save for empty slots
    for(Int i = 0; i < GetSaveCapacity(); i++)
    {
        if (!DoesSaveExist(i))
        {
            CreateSave(i);
        }
    }
}

};
